packages/ra-ui-materialui/src/input/AutocompleteInput/index.js

```diff
@@ -0,0 +1,396 @@
+ *    { id: 456, first_name: 'Jane', last_name: 'Austen' },
+ * ];
+ * const optionRenderer = choice => `${choice.first_name} ${choice.last_name}`;
+ * <AutocompleteInput source="author_id" choices={choices} optionText={optionRenderer} />
+ *
+ * The choices are translated by default, so you can use translation identifiers as choices:
+ * @example
+ * const choices = [
+ *    { id: 'M', name: 'myroot.gender.male' },
+ *    { id: 'F', name: 'myroot.gender.female' },
+ * ];
+ *
+ * However, in some cases (e.g. inside a `<ReferenceInput>`), you may not want
+ * the choice to be translated. In that case, set the `translateChoice` prop to false.
+ * @example
+ * <AutocompleteInput source="gender" choices={choices} translateChoice={false}/>
+ *
+ * The object passed as `options` props is passed to the material-ui <AutoComplete> component
+ *
+ * @example
+ * <AutocompleteInput source="author_id" options={{ fullWidth: true }} />
+ */
+const AutocompleteInput = ({
+    classes: classesOverride,
+    allowEmpty,
+    choices,
+    fullWidth,
+    helperText,
+    label,
+    limitChoicesToValue,
+    onBlur,
+    onChange,
+    onFocus,
+    options,
+    optionText,
+    optionValue,
+    resource,
+    setFilter,
+    shouldRenderSuggestions: shouldRenderSuggestionsOverride,
+    source,
+    suggestionComponent,
+    translateChoice,
+    validate,
+    ...rest
+}) => {
+    const translate = useTranslate();
+    const classes = useStyles({ classes: classesOverride });
+    const {
+        id,
+        input,
+        isRequired,
+        meta: { touched, error },
+    } = useInput({
+        onBlur,
+        onChange,
+        onFocus,
+        resource,
+        source,
+        type: 'checkbox',
+        validate,
+        ...rest,
+    });
+
+    let inputEl = useRef();
+    let anchorEl = useRef();
+
+    const updateFilter = useCallback(
+        value => {
+            if (setFilter) {
+                setFilter(value);
+            }
+        },
+        [setFilter]
+    );
+
+    useEffect(() => {
+        updateFilter('');
+    }, [input.value, updateFilter]);
+
+    const getSuggestionValue = suggestion => get(suggestion, optionValue);
+
+    const getSuggestionText = suggestion => {
+        if (!suggestion) return '';
+
+        const suggestionLabel =
+            typeof optionText === 'function'
+                ? optionText(suggestion)
+                : get(suggestion, optionText, '');
+
+        // We explicitly call toString here because AutoSuggest expect a string
+        return translateChoice
+            ? translate(suggestionLabel, { _: suggestionLabel }).toString()
+            : suggestionLabel.toString();
+    };
+
+    const getSuggestionTextFromValue = value => {
+        const currentChoice = choices.find(
+            choice => getSuggestionValue(choice) === value
+        );
+
+        return getSuggestionText(currentChoice);
+    };
+
+    const handleSuggestionSelected = suggestionText => {
```

Comment:
needs memoization with `useCallback`, just like most of the functions defined in this component